Add exponential backoff w/ jitter to healthz ping

AnOctopus opened the pull request:
Inludes a 60s max. Doubles each time, with 20% jitter.


## 📚 Context

Every web client pinging every 500ms is significant load during incidents in streamlit cloud.

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

Adds exponential backoff (with jitter, and a 60s max) to the frontend pings to `/healthz` for checking that it is still connected to the server.

  - [ ] This is a breaking API change
  - [ ] This is a visible (user-facing) change


## 🧪 Testing Done

- [ ] Screenshots included
- [x] Added/Updated unit tests
- [ ] Added/Updated e2e tests

AnOctopus committed:
Add exponential backoff w/ jitter to healthz ping

Inludes a 60s max. Doubles each time, with 20% jitter.

tconkling reviewed (changes requested):
Code looks good but needs a couple tests:
- test that the timeout is actually increasing between subsequent health checks, and maxes out
- test that the timeout starts at the minimum timeout each time in the loop

tconkling commented on frontend/src/lib/WebsocketConnection.tsx:
```diff
@@ -517,12 +522,22 @@ export function doHealthPing(
+  // because in some cases things fail very quickly, and all our fast retries
+  // end up bogging down the browser.
   const retry = (errorNode: React.ReactNode): void => {
     const tryDuration = (Date.now() - tryTimestamp) / 1000
```
It looks like `tryDuration` is in seconds as opposed to milliseconds - is this intentional?

AnOctopus replied:
No idea, that part was already in the code. It looks like a bug. That part of the function doesn't even feel very useful, since it only serves to make the ping interval not slowed down by the time it takes to do the ping, which is usually very low. I could fix it, to try and get minor responsiveness benefits, or I could remove it, because it currently isn't doing much of anything and nobody noticed, and have the function be a little simpler. I'm leaning towards removal, but I'm curious what you think

tconkling replied:
i think it's fine to remove it!

AnOctopus committed:
Merge branch 'develop' into feature/healthz-ping-exp-backoff-jitter

AnOctopus committed:
Remove mostly useless speedup in ping retries

This logic was supposed to have the ping timer progress even as we
waited for the ping request to resolve, but it took a milisecond
duration and converted it into seconds, then subtracted it from a
milisecond value, so it did almost nothing in practice. So it won't hurt
anything to remove to simplify the function.

AnOctopus committed:
Remove unused local variable in WebsocketConnection

AnOctopus committed:
Add test for values of websocket ping timeouts

AnOctopus committed:
Use constants in WebsocketConnection test

AnOctopus committed:
Add ping timeout reset test for WebsocketConnection

AnOctopus committed:
Healtcheck pings backoff independently per target

tconkling commented on frontend/src/lib/WebsocketConnection.tsx:
```diff
@@ -517,14 +524,19 @@ export function doHealthPing(
+      totalTries === 1
+        ? minimumTimeoutMs
+        : minimumTimeoutMs * 2 ** (totalTries - 1) * (1 + jitter)
+    // timeoutMs === 0 ? minimumTimeoutMs : timeoutMs * 2 * (1 + jitter)
```
Is this commented line intended to be here?

AnOctopus replied:
Nope. I'll remove it before I merge

tconkling reviewed (approved):
Looks great!

AnOctopus committed:
Remove spurious comment

AnOctopus committed:
Merge branch 'develop' into feature/healthz-ping-exp-backoff-jitter
